[doc] Develop ACC task. (#2969)
kubernetes-sigs/kueue: site/content/en/docs/tasks/dev/develop-acc.md

---
title: "Develop an AdmissionCheck Controller"
date: 2024-09-03
weight: 9
description: >
    Develop an AdmissionCheck Controller (ACC).
---

> An Admission Check Controller, sometime referred as **ACC**, is a component that manages the AdmissionChecks associated with it , by `spec.controllerName` and the workloads queued against ClusterQueues configured to use those AdmissionChecks.

Reviewer: ```suggestion
An Admission Check Controller, referred to as **ACC** in the rest of the doc,
is a component that manages the AdmissionChecks associated with it
(given by the corresponding value of `spec.controllerName`)
and the workloads queued against ClusterQueues configured to use those AdmissionChecks.
```

Author: done

> Check [Admission Check](/docs/concepts/admission_check/) section for the general description of the Admission Checks usage in Kueue.

Reviewer: ```suggestion
Read [Admission Check](/docs/concepts/admission_check/) to learn more about the mechanism from a user perspective.
```

## Subcomponents

The ACC can be built-in into Kueue or run in a different kubernetes controller manager and should implement reconcilers for AdmissionChecks and Workloads.

### AdmissionCheck Reconciler

> Monitors the AdmissionChecks in the cluster and maintain the `Active` condition of those that are associated with it (by `spec.controllerName`). Optionally it can watch custom [parameters](/docs/reference/kueue.v1beta1/#kueue-x-k8s-io-v1beta1-AdmissionCheckParametersReference) objects.

Reviewer: ```suggestion
Monitors the AdmissionChecks in the cluster and maintains the `Active` condition of those that are associated with it (by `spec.controllerName`).
Optionally, it can watch custom [parameters](/docs/reference/kueue.v1beta1/#kueue-x-k8s-io-v1beta1-AdmissionCheckParametersReference) objects.
```

The [Provisioning Admission Check Controller](/docs/admission-check-controllers/provisioning/) implements this in `pkg/controller/admissionchecks/provisioning/admissioncheck_reconciler.go`

### Workload Reconciler

> Is in charge of maintaining the [AdmissionCheckStates](/docs/concepts/admission_check/#admissioncheckstates) of individual workloads based on the ACC's custom logic being able to allow the admission of the workload, requeue or fail the workload. 

Reviewer: maybe something like this?

```suggestion
It is in charge of maintaining the [AdmissionCheckStates](/docs/concepts/admission_check/#admissioncheckstates) of individual workloads
based on the ACC's custom logic. It can either allow the admission, requeue or fail the workload. 
```

But I'm open to other suggestions.

The [Provisioning Admission Check Controller](/docs/admission-check-controllers/provisioning/) implements this in `pkg/controller/admissionchecks/provisioning/controller.go`

> 

Reviewer 2: AFAIK, additionally, the platform developer need to implement CustomResource and its controller used in the ACC parameters field.

So, could you mention that?

Author: done

> ## Util

Reviewer: ```suggestion
## Utility code
```

## Subcomponents

The ACC can be built-in into Kueue or run in a different kubernetes controller manager and should implement reconcilers for AdmissionChecks and Workloads.

### AdmissionCheck Reconciler

Monitors the AdmissionChecks in the cluster and maintains the `Active` condition of those that are associated with it (by `spec.controllerName`). 

Optionally, it can watch custom [parameters](/docs/reference/kueue.v1beta1/#kueue-x-k8s-io-v1beta1-AdmissionCheckParametersReference) objects.

The [Provisioning Admission Check Controller](/docs/admission-check-controllers/provisioning/) implements this in `pkg/controller/admissionchecks/provisioning/admissioncheck_reconciler.go`

### Workload Reconciler

It is in charge of maintaining the [AdmissionCheckStates](/docs/concepts/admission_check/#admissioncheckstates) of individual workloads
based on the ACC's custom logic. It can either allow the admission, requeue or fail the workload. 

The [Provisioning Admission Check Controller](/docs/admission-check-controllers/provisioning/) implements this in `pkg/controller/admissionchecks/provisioning/controller.go`

### Parameters Object Type

> Optionally, a custom cluster level object type holding the Admission Check specific parameters can be defined. 

Reviewer: ```suggestion
Optionally, you can define a cluster level object type to hold the Admission Check parameters specific to your implementation. 
```

Author: done

> Instances of this object type can be referenced in the AdmissionCheck definition by [spec.parameters](/docs/reference/kueue.v1beta1/#kueue-x-k8s-io-v1beta1-AdmissionCheckParametersReference) and provide implementation specific data to the ACC.

Reviewer: ```suggestion
Users can reference instances of this object type in the AdmissionCheck definition in [spec.parameters](/docs/reference/kueue.v1beta1/#kueue-x-k8s-io-v1beta1-AdmissionCheckParametersReference).
```

Author: done

> The [Provisioning Admission Check Controller](/docs/admission-check-controllers/provisioning/) uses [ProvisioningRequestConfig](/docs/reference/kueue.v1beta1/#kueue-x-k8s-io-v1beta1-ProvisioningRequestConfig) for this purpose.

Reviewer: ```suggestion
For example, the [Provisioning Admission Check Controller](/docs/admission-check-controllers/provisioning/) uses [ProvisioningRequestConfig](/docs/reference/kueue.v1beta1/#kueue-x-k8s-io-v1beta1-ProvisioningRequestConfig) for this purpose.
```

Author: done

## Utility Code

`pkg/util/admissioncheck` provides helper code that can be used to interact with AdmissionCheck's parameters and more.

## Examples

> Currently there are two ACCs implemented into kueue:

Author: > I assume developing an external controller is the target audience.

It is mentioned that it could be built-in or in a different controller L16.
